firebase-firestore/src/main/java/com/google/firebase/firestore/core/KeyFieldFilter.java

```diff
@@ -14,20 +14,25 @@
 
 package com.google.firebase.firestore.core;
 
+import static com.google.firebase.firestore.util.Assert.hardAssert;
+
 import com.google.firebase.firestore.model.Document;
+import com.google.firebase.firestore.model.DocumentKey;
 import com.google.firebase.firestore.model.FieldPath;
-import com.google.firebase.firestore.model.value.ReferenceValue;
+import com.google.firebase.firestore.model.value.ProtoValues;
+import com.google.firestore.v1.Value;
 
 /** Filter that matches on key fields (i.e. '__name__'). */
 public class KeyFieldFilter extends FieldFilter {
-  KeyFieldFilter(FieldPath field, Operator operator, ReferenceValue value) {
+  KeyFieldFilter(FieldPath field, Operator operator, Value value) {
     super(field, operator, value);
+    hardAssert(ProtoValues.isReferenceValue(value), "KeyFieldFilter expects a ReferenceValue");
   }
 
   @Override
   public boolean matches(Document doc) {
-    ReferenceValue referenceValue = (ReferenceValue) getValue();
-    int comparator = doc.getKey().compareTo(referenceValue.getKey());
+    DocumentKey referencedKey = DocumentKey.fromName(getValue().getReferenceValue());
```

Comment:
This seems like something we could store as a field. It doesn't change for every document we match.

Reply:
Done.